File: frontend/src/api/extractionReview.ts
/**
 * API client for extraction review endpoints.
 *
 * Handles human-in-the-loop review of LLM extractions.
 */
import { apiFetch } from "./client";
import {
  appendOptionalNumber,
  appendOptionalParam,
  buildQueryString,
  createSearchParams,
} from "./queryString";
import type { ExtractedEntity, ExtractedRelation } from "../types/extraction";

export type ReviewDecision = "approve" | "reject";
export type ExtractionStatus = "auto_verified" | "pending" | "approved" | "rejected";
export type ExtractionType = "entity" | "relation";
export type StagedExtractionData = ExtractedEntity | ExtractedRelation;

export interface StagedExtractionRead {
  id: string;
  extraction_type: ExtractionType;
  status: ExtractionStatus;
  source_id: string;
  extraction_data: StagedExtractionData;
  validation_score: number;
  confidence_adjustment: number;
  validation_flags: string[];
  matched_span?: string | null;
  materialized_entity_id?: string;
  materialized_relation_id?: string;
  reviewed_by?: string;
  reviewed_at?: string;
  review_notes?: string;
  llm_model?: string;
  llm_provider?: string;
  auto_commit_eligible: boolean;
  auto_commit_threshold?: number | null;
  auto_approved: boolean;
  created_at: string;
}

export interface StagedExtractionListResponse {
  extractions: StagedExtractionRead[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface ReviewStats {
  total_pending: number;
  total_approved: number;
  total_rejected: number;
  total_auto_verified: number;
  pending_entities: number;
  pending_relations: number;
  avg_validation_score: number;
  high_confidence_count: number;
  flagged_count: number;
}

export interface ReviewDecisionRequest {
  decision: ReviewDecision;
  notes?: string;
}

export interface BatchReviewRequest {
  extraction_ids: string[];
  decision: ReviewDecision;
  notes?: string;
}

export interface BatchReviewResponse {
  total_requested: number;
  succeeded: number;
  failed: number;
  failed_ids: string[];
  materialized_entities: string[];
  materialized_relations: string[];
}

export interface MaterializationResult {
  success: boolean;
  extraction_id: string;
  extraction_type: ExtractionType;
  materialized_entity_id?: string | null;
  materialized_relation_id?: string | null;
  error?: string | null;
}

export interface EntityMergeCandidateEntity {
  id: string;
  slug: string;
  summary?: Record<string, string> | null;
}

export interface StagedEntityMergeCandidate {
  target: EntityMergeCandidateEntity;
  similarity: number;
  reason: string;
  score_factors?: Record<string, string | number | boolean>;
  proposed_action: "approve_then_merge";
}

export interface StagedExtractionFilters {
  status?: ExtractionStatus;
  extraction_type?: ExtractionType;
  source_id?: string;
  min_validation_score?: number;
  max_validation_score?: number;
  has_flags?: boolean;
  auto_commit_eligible?: boolean;
  page?: number;
  page_size?: number;
  sort_by?: "created_at" | "validation_score" | "confidence_adjustment";
  sort_order?: "asc" | "desc";
}

/**
 * List pending extractions that need review.
 *
 * GET /api/extraction-review/pending
 */
export async function listPendingExtractions(
  filters?: StagedExtractionFilters
): Promise<StagedExtractionListResponse> {
  const params = createSearchParams((query) => {
    appendOptionalParam(query, "extraction_type", filters?.extraction_type);
    appendOptionalParam(query, "source_id", filters?.source_id);
    appendOptionalNumber(query, "min_validation_score", filters?.min_validation_score);
    appendOptionalNumber(query, "max_validation_score", filters?.max_validation_score);
    appendOptionalParam(query, "has_flags", filters?.has_flags);
    appendOptionalParam(query, "auto_commit_eligible", filters?.auto_commit_eligible);
    appendOptionalNumber(query, "page", filters?.page);
    appendOptionalNumber(query, "page_size", filters?.page_size);
    appendOptionalParam(query, "sort_by", filters?.sort_by);
    appendOptionalParam(query, "sort_order", filters?.sort_order);
  });

  return apiFetch<StagedExtractionListResponse>(
    `/extraction-review/pending${buildQueryString(params)}`
  );
}

/**
 * Get review statistics.
 *
 * GET /api/extraction-review/stats
 */
export async function getReviewStats(): Promise<ReviewStats> {
  return apiFetch<ReviewStats>("/extraction-review/stats");
}

/**
 * Get a single staged extraction by ID.
 *
 * GET /api/extraction-review/{id}
 */
export async function getStagedExtraction(
  extractionId: string
): Promise<StagedExtractionRead> {
  return apiFetch<StagedExtractionRead>(`/extraction-review/${extractionId}`);
}

/**
 * List existing entities that may match one staged entity extraction.
 *
 * GET /api/extraction-review/{id}/entity-merge-candidates
 */
export async function listStagedEntityMergeCandidates(
  extractionId: string
): Promise<StagedEntityMergeCandidate[]> {
  return apiFetch<StagedEntityMergeCandidate[]>(
    `/extraction-review/${extractionId}/entity-merge-candidates`
  );
}

/**
 * Review an extraction (approve or reject).
 *
 * POST /api/extraction-review/{id}/review
 */
export async function reviewExtraction(
  extractionId: string,
  request: ReviewDecisionRequest
): Promise<MaterializationResult> {
  return apiFetch<MaterializationResult>(
    `/extraction-review/${extractionId}/review`,
    {
      method: "POST",
      body: JSON.stringify(request),
    }
  );
}

/**
 * Correct the relation_type of a pending staged relation extraction.
 *
 * PATCH /api/extraction-review/{id}/relation-type
 */
export async function correctRelationType(
  extractionId: string,
  relationType: string,
): Promise<StagedExtractionRead> {
  return apiFetch<StagedExtractionRead>(
    `/extraction-review/${extractionId}/relation-type`,
    {
      method: "PATCH",
      body: JSON.stringify({ relation_type: relationType }),
    }
  );
}

/**
 * Batch review multiple extractions.
 *
 * POST /api/extraction-review/batch-review
 */
export async function batchReview(
  request: BatchReviewRequest
): Promise<BatchReviewResponse> {
  return apiFetch<BatchReviewResponse>("/extraction-review/batch-review", {
    method: "POST",
    body: JSON.stringify(request),
  });
}

/**
 * List all extractions with filtering (admin only).
 *
 * GET /api/extraction-review/all
 */
export async function listAllExtractions(
  filters?: StagedExtractionFilters
): Promise<StagedExtractionListResponse> {
  const params = createSearchParams((query) => {
    appendOptionalParam(query, "status", filters?.status);
    appendOptionalParam(query, "extraction_type", filters?.extraction_type);
    appendOptionalParam(query, "source_id", filters?.source_id);
    appendOptionalNumber(query, "min_validation_score", filters?.min_validation_score);
    appendOptionalNumber(query, "max_validation_score", filters?.max_validation_score);
    appendOptionalParam(query, "has_flags", filters?.has_flags);
    appendOptionalParam(query, "auto_commit_eligible", filters?.auto_commit_eligible);
    appendOptionalNumber(query, "page", filters?.page);
    appendOptionalNumber(query, "page_size", filters?.page_size);
    appendOptionalParam(query, "sort_by", filters?.sort_by);
    appendOptionalParam(query, "sort_order", filters?.sort_order);
  });

  return apiFetch<StagedExtractionListResponse>(
    `/extraction-review/all${buildQueryString(params)}`
  );
}

File: backend/app/api/extraction_review.py
"""
API endpoints for extraction review workflow.

Provides human-in-the-loop review interface for staged LLM extractions.
"""
import logging
from fastapi import APIRouter, Depends, Query, status
from uuid import UUID
from pydantic import BaseModel

from app.api.service_dependencies import get_extraction_review_service
from app.dependencies.auth import get_current_active_superuser
from app.models.user import User
from app.schemas.staged_extraction import (
    StagedExtractionRead,
    StagedExtractionListResponse,
    ReviewDecisionRequest,
    BatchReviewRequest,
    AutoCommitResponse,
    BatchReviewResponse,
    ReviewStats,
    MaterializationResult,
    StagedExtractionFilters,
)
from app.services.extraction_review_service import ExtractionReviewService
from app.utils.errors import AppException, ErrorCode

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/extraction-review", tags=["extraction-review"])


# =============================================================================
# Listing and Statistics
# =============================================================================

@router.get("/pending", response_model=StagedExtractionListResponse)
async def list_pending_extractions(
    filters: StagedExtractionFilters = Depends(),
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    List pending extractions awaiting review.

    Supports filtering by type, validation score, and flag presence.
    """
    pending_filters = filters.model_copy(update={"status": "pending"})

    extractions, total = await service.list_extractions(pending_filters)

    return StagedExtractionListResponse(
        extractions=[StagedExtractionRead.model_validate(e) for e in extractions],
        total=total,
        page=pending_filters.page,
        page_size=pending_filters.page_size,
        has_more=(pending_filters.page * pending_filters.page_size) < total,
    )


@router.get("/stats", response_model=ReviewStats)
async def get_review_stats(
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    Get statistics about staged extractions.

    Includes counts by status, type, and quality metrics.
    """
    return await service.get_stats()


@router.get("/all", response_model=StagedExtractionListResponse)
async def list_all_extractions(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=50, ge=1, le=100),
    status: str | None = None,
    extraction_type: str | None = None,
    source_id: UUID | None = None,
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    List all staged extractions with filtering.

    Admin only. Allows viewing extractions in any status.
    """
    filters = StagedExtractionFilters(
        status=status,
        extraction_type=extraction_type,
        source_id=source_id,
        page=page,
        page_size=page_size,
    )

    extractions, total = await service.list_extractions(filters)

    return StagedExtractionListResponse(
        extractions=[StagedExtractionRead.model_validate(e) for e in extractions],
        total=total,
        page=page,
        page_size=page_size,
        has_more=(page * page_size) < total,
    )


@router.get("/{extraction_id}", response_model=StagedExtractionRead)
async def get_extraction(
    extraction_id: UUID,
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    Get details of a specific staged extraction.
    """
    extraction = await service.get_extraction(extraction_id)

    if not extraction:
        raise AppException(
            status_code=404,
            error_code=ErrorCode.NOT_FOUND,
            message="Staged extraction not found",
            details=f"Staged extraction with ID '{extraction_id}' does not exist",
            context={"extraction_id": str(extraction_id)}
        )

    return StagedExtractionRead.model_validate(extraction)


class RelationTypeCorrectionRequest(BaseModel):
    relation_type: str


@router.patch("/{extraction_id}/relation-type", response_model=StagedExtractionRead)
async def correct_relation_type(
    extraction_id: UUID,
    request: RelationTypeCorrectionRequest,
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    Correct the relation_type of a pending staged relation extraction.

    Updates extraction_data.relation_type in-place. Intended for curators
    who want to reassign an 'other' relation (or a model-invented type) to
    an existing controlled-vocabulary type before approving.
    """
    extraction = await service.get_extraction(extraction_id)
    if not extraction:
        raise AppException(
            status_code=404,
            error_code=ErrorCode.NOT_FOUND,
            message="Staged extraction not found",
            context={"extraction_id": str(extraction_id)},
        )
    if extraction.extraction_type != "relation":
        raise AppException(
            status_code=400,
            error_code=ErrorCode.VALIDATION_ERROR,
            message="relation_type correction only applies to relation extractions",
        )

    data = dict(extraction.extraction_data or {})
    data["relation_type"] = request.relation_type
    extraction.extraction_data = data
    await service.db.commit()
    await service.db.refresh(extraction)
    return StagedExtractionRead.model_validate(extraction)


# =============================================================================
# Review Actions
# =============================================================================

@router.post("/{extraction_id}/review", response_model=MaterializationResult)
async def review_extraction(
    extraction_id: UUID,
    decision: ReviewDecisionRequest,
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    Approve or reject a staged extraction.

    If approved, automatically materializes the extraction into the knowledge graph.
    """
    if decision.decision == "approve":
        result = await service.approve_extraction(
            extraction_id=extraction_id,
            reviewer_id=current_user.id,
            notes=decision.notes,
            auto_materialize=True,
        )
        return result
    else:  # reject
        staged = await service.get_extraction(extraction_id)
        if not staged:
            raise AppException(
                status_code=404,
                error_code=ErrorCode.NOT_FOUND,
                message="Staged extraction not found",
                details=f"Staged extraction with ID '{extraction_id}' does not exist",
                context={"extraction_id": str(extraction_id)}
            )
        success = await service.reject_extraction(
            extraction_id=extraction_id,
            reviewer_id=current_user.id,
            notes=decision.notes,
        )
        if success:
            return MaterializationResult(
                success=True,
                extraction_id=extraction_id,
                extraction_type=staged.extraction_type.value,
            )
        else:
            raise AppException(
                status_code=400,
                error_code=ErrorCode.VALIDATION_ERROR,
                message="Failed to reject extraction",
                details="The extraction could not be rejected",
                context={"extraction_id": str(extraction_id)}
            )


@router.post("/batch-review", response_model=BatchReviewResponse)
async def batch_review_extractions(
    request: BatchReviewRequest,
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    Review multiple extractions at once.

    All extractions will receive the same decision and notes.
    """
    result = await service.batch_review(
        extraction_ids=request.extraction_ids,
        decision=request.decision,
        reviewer_id=current_user.id,
        notes=request.notes,
    )

    return result


# =============================================================================
# Auto-Commit Management
# =============================================================================

@router.post("/auto-commit", response_model=AutoCommitResponse)
async def trigger_auto_commit(
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    Manually trigger auto-commit of all eligible pending extractions.

    Admin only. This will approve and materialize all high-confidence extractions
    that meet auto-commit criteria.
    """
    return await service.auto_commit_eligible_extractions()


@router.delete("/{extraction_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_staged_extraction(
    extraction_id: UUID,
    service: ExtractionReviewService = Depends(get_extraction_review_service),
    current_user: User = Depends(get_current_active_superuser),
):
    """
    Delete a staged extraction.

    Admin only. Useful for removing erroneous or duplicate extractions.
    """
    deleted = await service.delete_extraction(extraction_id)
    if not deleted:
        raise AppException(
            status_code=404,
            error_code=ErrorCode.NOT_FOUND,
            message="Staged extraction not found",
            details=f"Staged extraction with ID '{extraction_id}' does not exist",
            context={"extraction_id": str(extraction_id)}
        )

    return None
